extensions-contrib/opentelemetry-extensions/src/main/java/org/apache/druid/data/input/opentelemetry/protobuf/OpenTelemetryMetricsProtobufReader.java

```diff
@@ -0,0 +1,222 @@
+import io.opentelemetry.proto.metrics.v1.NumberDataPoint;
+import org.apache.druid.data.input.InputEntityReader;
+import org.apache.druid.data.input.InputRow;
+import org.apache.druid.data.input.InputRowListPlusRawValues;
+import org.apache.druid.data.input.MapBasedInputRow;
+import org.apache.druid.data.input.impl.ByteEntity;
+import org.apache.druid.data.input.impl.DimensionsSpec;
+import org.apache.druid.java.util.common.CloseableIterators;
+import org.apache.druid.java.util.common.parsers.CloseableIterator;
+import org.apache.druid.java.util.common.parsers.ParseException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+public class OpenTelemetryMetricsProtobufReader implements InputEntityReader
+{
+
+  private final ByteEntity source;
+  private final String metricDimension;
+  private final String valueDimension;
+  private final String metricAttributePrefix;
+  private final String resourceAttributePrefix;
+  private final Predicate<String> attributeMembership;
+  private final List<String> schemaDimensions;
+
+  public OpenTelemetryMetricsProtobufReader(
+      DimensionsSpec dimensionsSpec,
+      ByteEntity source,
+      String metricDimension,
+      String valueDimension,
+      String metricLabelPrefix,
+      String resourceLabelPrefix
+  )
+  {
+    this.source = source;
+    this.metricDimension = metricDimension;
+    this.valueDimension = valueDimension;
+    this.metricAttributePrefix = metricLabelPrefix;
+    this.resourceAttributePrefix = resourceLabelPrefix;
+
+    this.schemaDimensions = dimensionsSpec.getDimensionNames();
+    if (!schemaDimensions.isEmpty()) {
+      this.attributeMembership = schemaDimensions::contains;
+    } else {
+      this.attributeMembership = att -> !dimensionsSpec.getDimensionExclusions().contains(att);
+    }
+  }
+
+  @Override
+  public CloseableIterator<InputRow> read()
+  {
+    return CloseableIterators.withEmptyBaggage(readAsList().iterator());
+  }
+
+  List<InputRow> readAsList()
+  {
+    try {
+      return parseMetricsData(MetricsData.parseFrom(source.getBuffer()));
+    }
+    catch (InvalidProtocolBufferException e) {
+      throw new ParseException(e, "Protobuf message could not be parsed");
+    }
+  }
+
+  private List<InputRow> parseMetricsData(final MetricsData metricsData)
+  {
+    return metricsData.getResourceMetricsList()
+            .stream()
+            .flatMap(resourceMetrics -> resourceMetrics.getInstrumentationLibraryMetricsList()
+                    .stream()
+                    .flatMap(libraryMetrics -> libraryMetrics.getMetricsList()
+                            .stream()
+                            .flatMap(metric -> parseMetric(metric, resourceMetrics.getResource().getAttributesList())
+                                    .stream())))
+            .collect(Collectors.toList());
```

Comment:
we should check to see what the performance impact of using the streaming collections is here. For the OpenCensus format, we switched to foreach loops, which significantly improved performance

Reply:
The code seems to be a good bit faster than the OpenCensus InputFormat for similar loads.
<img width="617" alt="OCensusBenchmark" src="https://user-images.githubusercontent.com/16495842/151469633-205d6f99-feee-4fee-a50a-1299c0f6ad47.png">

<img width="759" alt="Screen Shot 2022-01-27 at 4 21 30 PM" src="https://user-images.githubusercontent.com/16495842/151469667-f814a658-5b25-4350-ad7c-cc45e51d8223.png">